broadinstitute/single_cell_portal_core
# Fast ontology validation for AnnData (SCP-5800)

This drastically speeds up validation of controlled terms in AnnData files.  That makes upload smoother, and thus faster.

### Impact
[Previously](https://github.com/broadinstitute/single_cell_portal_core/pull/2122), headers and some content were checked right away, but the ontology IDs and labels in H5AD files were validated only after uploading.  Uploading from a local machine often takes 10-30 minutes, and validation errors would be emailed to the author.  

Now, ontology errors are reported in 5-10 seconds, right in the app.  That's dramatically faster and clearer for [18-50% of AnnData validation failures](https://mixpanel.com/project/2120588/view/19059/app/boards#id=734187&editor-card-id=%22report-30146183%22).  This better UX will help authors add more studies and, as we mature integration, engage with useful tools like [Cell Annotation Service](https://singlecell.broadinstitute.org/single_cell/features/2024-09-18-explore-cell-typing-with-ai-take-a-sneak-peek-at-cellarium-cell-annotation-service) (CAS).

### Demo
Here's how it looks!

https://github.com/user-attachments/assets/89543d58-2dca-4c61-a86f-ed13bdb22505

### Technical overview
Two problems make ontology validation more complex than other validation rules:
1. It required using a third-party remote API
2. Collating IDs and labels -- to check that the given label is correct for the given ID -- is not trivial to do in a streaming manner for AnnData.  Streaming is needed for efficient memory use, which is critical when operating in a web browser client.

More context on solving problem 2 is in a comment beginning "AnnData organizes" in `validate-anndata.js`.

The diagrams below outline implementation details for solving problem 1.  These complement the description of related backend work in https://github.com/broadinstitute/scp-ingest-pipeline/pull/364.

#### Previous implementation: validate after upload
Historically, after upload, our server-side Ingest Pipeline would do ontology validation by querying EBI OLS, a remote API.

<img width="957" alt="AnnData_ontology_validation_before__SCP_2024-09-23" src="https://github.com/user-attachments/assets/5a596f10-571c-4cd6-834b-90f21c49cc3e">

#### New implementation: validate before upload
Now, ontology validation is done _before_ upload, using minified ontologies.  Server-side validation is still also done for robustness.  The big value-add here is that failures happen fast -- before upload.

<img width="959" alt="AnnData_ontology_validation_now__SCP_2024-09-23" src="https://github.com/user-attachments/assets/b02d8cc6-97e1-4d19-b1cf-cf4a3eb3d223">

#### Related backend improvement: minify ontologies
The prior implementation queried EBI OLS via the server at runtime.  That was possible in the new client-side implementation, but would have been less reliable and not as snappy as a tailored approach.  

Instead of querying OLS, this new implementation queries slim ontologies containing only the needed reference data.  Ontologies are minified at "build time", when any PR is opened in Ingest Pipeline.  The slim ontologies are then loaded onto the client at runtime, when an AnnData file is selected in the upload UI.  More backend details are in https://github.com/broadinstitute/scp-ingest-pipeline/pull/364 and https://github.com/broadinstitute/scp-ingest-pipeline/pull/365.

<img width="911" alt="Minify_ontologies_for_fast_validation__SCP_2024-09-23" src="https://github.com/user-attachments/assets/88f46250-ecee-440e-a344-af2a293966a8">

#### Instant reference retrieval: fetch ontologies only once
The above approach makes getting ontology reference data fast -- a little over 2 seconds.  To get that down to 0.2 seconds, this new work caches parsed mini-ontologies on the client.  

On a user's very first page load with this new code, the app fetch ontologies and decompresses them on the client.  This gunzip is done in JS because the files are gzipped at rest, not merely server-transcoded.  It then puts the resulting TSVs in the browser's [service worker cache](https://web.dev/articles/service-worker-caching-and-http-caching).  Subsequent page loads only check for updates to a `version.txt` file, and instantly load decompressed data from local disk.  Remote ontology references are then only re-fetched when `version.txt` occasionally updates.

This last optimization saves ~2 seconds in overall AnnData CSFV -- 7-12 seconds before, 5-10 seconds after.  For local development, iteration times speed up from 5 seconds before to 3 seconds after.

<img width="958" alt="Fetch_ontologies_only_once__SCP_2024-09-23" src="https://github.com/user-attachments/assets/7af80b39-4068-4b95-aa32-5d2d63d07b5d">

### Next steps
This novel approach could be enhanced.

1.  **Extend ontology CSFV to classic files**.  Flat files / TSVs still account for a huge portion of uploads, and ontology validation errors remain common.  More specifically, [between 36% and 59% of validation errors in classic files](https://mixpanel.com/project/2120588/view/19059/app/boards#id=734187&editor-card-id=%22report-30146183%22) remain frustratingly slow-to-detect.  Extending this new mini-ontology work could speed up those author iteration times from 5-10 minutes to 2-5 seconds.

2. **Apply AnnData ontology CSFV to remote files**.  While manually testing a 6.7 GB AnnData file, I encountered the most surprising bug of the year for me.  It seems the `hdf5-indexed-reader` library produces _incorrect_ data in limited circumstances for remote AnnData.  That yields false-positive ontology validation failures for big remote files.  More detail is in SCP-5813.  In summary, we can try to fix that underlying library bug, or failing that report the issue upstream with relevant findings from our attempted fix.  For now I've disabled ontology validation for remote AnnData files, e.g. "Use bucket path".  (I confirmed this does not affect AnnData ontology CSFV for local files, which is the main beneficiary for this PR.)

### Test
Automated tests verify some updated behavior.  Optionally, if you'd like to manually test:

1.  Go to AnnData CSFV study
2. Select file "invalid_disease_label.h5ad"
3. Confirm error message appears
4. Upload that file to your study bucket
5. Click "Use bucket path", paste invalid_disease_label.h5ad into text input
6. Click outside text input
7. Confirm error message appears
8. Select valid file from local machine
9. Confirm no error message appears
10. Select valid file from bucket
11. Confirm no error message appears

### Further details
More context is available in [2024-09-17 SCP demo video](https://drive.google.com/file/d/1hqiQkwGWSM52RYIExnmDu3ILC_-dwoCc/view?t=42s).  This satisfies SCP-5800.
